p2p/src/interface/p2p_interface.rs

```diff
@@ -33,3 +32,8 @@ pub trait P2pInterface: Send + Sync {
 
     async fn submit_transaction(&mut self, tx: SignedTransaction) -> crate::Result<()>;
 }
+
+#[async_trait::async_trait]
+pub trait P2pSubsystemInterface: 'static {
+    async fn run(self, call: CallRequest<dyn P2pInterface>, shutdown: ShutdownRequest);
+}
```

Comment:
Looks like it's not that easy. I guess no need to spend time on it. Sorry!

Reply:
@pavel-kokolemin Don't worry, the current solution isn't really optimal.

@TheQuantumPhysicist We discussed this with Pavel and he proposed a slightly different solution: to extend the p2p interface with the shutdown method. Something like this:
```rust
trait P2pInterface: Send + Sync {
    async fn do_shutdown(self);
}

#[async_trait::async_trait]
impl subsystem::Subsystem for Box<dyn P2pInterface> {
    async fn shutdown(self) {
        self.do_shutdown().await;
    }
}
```
I think it is somewhat better, but I still don't like the idea of the shutdown method in the interface.

Reply:
We shouldn't have a shutdown method in the p2p interface... that's just wrong. Shutdown events should strictly come from subsystem. Now I have to say I'm not familiar with the current predicament so I don't understand why this is even being discussed.

Reply:
I guess what I'm asking is: Why are we not using concrete methods and are using interfaces instead for run and shutdown?!

Reply:
The `P2p` struct is generic over the networking service (transport) and the concrete transport is chosen based on runtime parameters, so it is impossible to return it from a function:
```rust
pub async fn make_p2p<S: PeerDbStorage + 'static>(
    chain_config: Arc<ChainConfig>,
    p2p_config: Arc<P2pConfig>,
    chainstate_handle: subsystem::Handle<Box<dyn chainstate_interface::ChainstateInterface>>,
    mempool_handle: MempoolHandle,
    time_getter: TimeGetter,
    peerdb_storage: S,
) -> Result<Box<dyn P2pInterface>> {
    if let Some(true) = p2p_config.disable_noise {
        let transport = make_p2p_transport_unencrypted();

        let p2p = P2p::<P2pNetworkingServiceUnencrypted>::new(
            transport,
            bind_addresses,
            chain_config,
            p2p_config,
            chainstate_handle,
            mempool_handle,
            time_getter,
            peerdb_storage,
        )
        .await?;

        Ok(Box::new(p2p))
    } else if let Some(socks5_proxy) = &p2p_config.socks5_proxy {
        let transport = make_p2p_transport_socks5_proxy(socks5_proxy);

        let p2p = P2p::<P2pNetworkingServiceSocks5Proxy>::new(
            transport,
            bind_addresses,
            chain_config,
            p2p_config,
            chainstate_handle,
            mempool_handle,
            time_getter,
            peerdb_storage,
        )
        .await?;

        Ok(Box::new(p2p))
    } else {
        let transport = make_p2p_transport();

        let p2p = P2p::<P2pNetworkingService>::new(
            transport,
            bind_addresses,
            chain_config,
            p2p_config,
            chainstate_handle,
            mempool_handle,
            time_getter,
            peerdb_storage,
        )
        .await?;

        Ok(Box::new(p2p))
    }
}
```

Therefore the interface (`Box<dyn P2pInterface>`) is returned. As a result, the `Subsystem` trait is implemented for the interface instead of the `P2p` type.

Reply:
So in a call we clarified that this trait isn't necessary, and the implementee struct can just be returned instead.

I hope the same would be done for the mempool. I just realized it's done that way there as well.